src/nifgen/templates/write_waveform.py.mako

```diff
@@ -0,0 +1,34 @@
+<%page args="f, config"/>\
+<%
+    '''Dispatches to the appropriate "write waveform" method based on the waveform type.'''
+    import build.helper as helper
+%>\
+    def ${f['python_name']}(${helper.get_params_snippet(f, helper.ParameterUsageOptions.SESSION_METHOD_DECLARATION)}):
+        '''${f['python_name']}
+
+        ${helper.get_function_docstring(f['name'], config, indent=8)}
+        '''
+        try:
+            import numpy
+            numpy_imported = True
+        except ImportError:
+            numpy_imported = False
+
+        if numpy_imported is True and type(data) == numpy.ndarray:
+            if data.dtype == numpy.float64:
+                write_named_method = self._write_named_waveform_f64_numpy
+                write_handle_method = self._write_waveform_numpy
+            elif data.dtype == numpy.int16:
+                write_named_method = self._write_named_waveform_i16_numpy
+                write_handle_method = self._write_binary16_waveform_numpy
+            else:
+                raise TypeError("Unsupported dtype. Is {0}, expected {1} or {2}".format(data.dtype, numpy.float64, numpy.int16))
+        else:
+            write_named_method = self._write_named_waveform_f64
```

Comment:
What if `type(data)` is a numpy type but you failed importing numpy? You then assume it's a list of floats so you should assert that the type here is list of floats.

Reply:
I'm not sure how that would happen. If the client has a numpy type it means the client successfully imported numpy. I don't know under which circumstances  other than the system being really hosed this would fail importing numpy.

That said, the type used in _write_named_waveform_f64 does **not** have to be a list of floats. Any iterable should work fine.

Reply:
Okay.